src/mers/kmer.jl

```diff
@@ -0,0 +1,324 @@
+@inline capacity(::Type{Kmer{A,K,N}}) where {A,K,N} = div(64N, 2)
+@inline capacity(seq::Kmer) = capacity(typeof(seq))
+@inline n_unused(::Type{Kmer{A,K,N}}) where {A,K,N} = capacity(Kmer{A,K,N}) - K
+@inline n_unused(seq::Kmer) = n_unused(typeof(seq))
+
+@inline function checkmer(::Type{Kmer{A,K,N}}) where {A,K,N}
+    if K < 1
+        throw(ArgumentError("Bad kmer parameterisation. K must be greater than 0."))
+    end
+    n = ifelse(rem(2K, 64) != 0, div(2K, 64) + 1, div(2K, 64))
+    if n !== N
+        # This has been significantly changed conceptually from before. Now we
+        # don't just check K, but *enforce* the most appropriate N for K.
+        throw(ArgumentError("Bad kmer parameterisation. For K = $K, N should be $n"))
+    end
+end
+
+@inline Base.length(x::Kmer{A,K,N}) where {A,K,N} = K
+@inline Base.summary(x::Kmer{DNAAlphabet{2},K,N}) where {K,N} = string("DNA ", K, "-mer")
+@inline Base.summary(x::Kmer{RNAAlphabet{2},K,N}) where {K,N} = string("RNA ", K, "-mer")
+
+function Base.typemin(::Type{Kmer{A,K,N}}) where {A,K,N}
+    #checkmer(Kmer{A,K,N}) Got moved to constructor.
+    return Kmer{A,K,N}(ntuple(i -> zero(UInt64), N))
+end
+
+function Base.typemax(::Type{Kmer{A,K,N}}) where {A,K,N}
+    #checkmer(Kmer{A,K,N}) Got moved to constructor.
+    return Kmer{A,K,N}((typemax(UInt64) >> (64N - 2K), ntuple(i -> typemax(UInt64), N - 1)...))
+end
+
+function Base.rand(::Type{Kmer{A,K,N}}) where {A,K,N}
+    #checkmer(Kmer{A,K,N}) Got moved to constructor.
+    return Kmer{A,K,N}((rand(UInt64) >> (64N - 2K), ntuple(i -> rand(UInt64), N - 1)...))
+end
+
+Base.rand(::Type{Kmer{A,K}}) where {A,K} = rand(kmertype(Kmer{A,K}))
+
+function Base.rand(::Type{T}, size::Integer) where {T<:Kmer}
+    return [rand(T) for _ in 1:size]
+end
+
+###
+### Old Mer Base Functions - not transferred to new type.
+###
+#@inline encoded_data_type(::Type{Mer{A,K}}) where {A,K} = UInt64
+#@inline encoded_data_type(::Type{BigMer{A,K}}) where {A,K} = UInt128
+#@inline encoded_data_type(x::AbstractMer) = encoded_data_type(typeof(x))
+#@inline encoded_data(x::AbstractMer) = reinterpret(encoded_data_type(typeof(x)), x)
+#@inline ksize(::Type{T}) where {A,K,T<:AbstractMer{A,K}} = K
+#@inline Base.unsigned(x::AbstractMer) = encoded_data(x)
+#Base.:-(x::AbstractMer, y::Integer) = typeof(x)(encoded_data(x) - y % encoded_data_type(x))
+#Base.:+(x::AbstractMer, y::Integer) = typeof(x)(encoded_data(x) + y % encoded_data_type(x))
+#Base.:+(x::AbstractMer, y::AbstractMer) = y + x
+#Alphabet(::Type{Mer{A,K} where A<:NucleicAcidAlphabet{2}}) where {K} = Any
+
+@inline function inbounds_getindex(x::Kmer{A,K,N}, i::Integer) where {A,K,N}
+    # Emulation of BitIndex type
+    i′ = i + div(64N - 2K, 2)
+    val = (i′ - 1) << 1
+    idx = (val >> 6) + 1
+    off = 62 - (val & (UInt8(64) - 0x01))
+    @inbounds begin
+        chunk = x.data[idx]
+    end
+    bits = (chunk >> off) & UInt64(3)
+    return reinterpret(eltype(x), 0x01 << bits)
+end
+
```

Comment:
Speed and laziness on my part basically :P, I will massage it into the proper method hierarchy.

Reply:
Ok, this has been incorporated into the lastest PR.

Unlike LongSequence, Kmers store bases in chunks from MSB -> LSB.
This means their "unused bits" are in the MSBs of the head before the first base (cf. the unused space in the LSBs of the tail for LongSequence).

So to deal with that:

* Kmer defines a specialisation of `bitindex(seq::Kmer, i::Integer)`, that basically adds to `i` how much-unused space in In the head (in bases), before constructing the bitindex - reminiscent of how LongSequence had to work out what `i` meant in the context of subsequences etc, before constructing the BitIndex.

* I've added a specialisation to `extract_encoded_element(::BitIndex, ::NTuple)` It behaves as for vectors, except that since Kmers order bases from MSB -> LSB, it adjusts the offset slightly before pulling out the bits and passing them to decode. Doing it this way is almost line for line of ASM equivalent to the `inbounds_getindex`  for Kmers that is now removed. Long term I think perhaps there should be a LSB or MSB trait for the BitIndex that specialises calculation of offsets and so on, but this is fine for now if we assume sequences using tuples just store in MSB order - which so far is only kmers. There may be a perf hit as unlike the kmer specific inbounds_getindex, `decode` introduces a branch that may throw a `DecodeError`. But this way it's like all other sequences, and if we decide decoding needs to be revisited to avoid throwing, or adding `unsafe_decode` or some such thing, we could in the future.